packages/commonwealth/client/scripts/controllers/server/sessions.ts

```diff
@@ -117,34 +184,69 @@ class SessionsController {
     // `chainBaseToCanvasChainId` replaces idOrPrefix with the appropriate
     // chainID for non-eth, non-cosmos chains.
     const idOrPrefix =
-      chainBase == ChainBase.CosmosSDK
+      chainBase === ChainBase.CosmosSDK
         ? app.chain?.meta.bech32Prefix
         : app.chain?.meta.node?.ethChainId;
-    const chainId = chainBaseToCanvasChainId(chainBase, idOrPrefix);
+    const canvasChainId = chainBaseToCanvasChainId(chainBase, idOrPrefix);
 
     // Try to request a new session from the user, if one was not found.
     const controller = this.getSessionController(chainBase);
 
     // Load any past session
     const hasAuthenticatedSession = await controller.hasAuthenticatedSession(
-      chainId
+      canvasChainId,
+      address
     );
 
-    // TODO: Turn on the session sign-in modal
+    // Get a new session signature.
     if (!hasAuthenticatedSession) {
-      return {
-        session: JSON.stringify({}),
-        action: JSON.stringify({}),
-        hash: '0x0000000000000000000000000000000000000000000000000000000000000000',
-      };
-      // await sessionSigninModal().catch((err) => {
-      //   console.log('Login failed');
-      //   throw err;
-      // });
+      const matchingAccount = app.user.addresses.find(
+        (a) => a.address === address
+      );
+      const walletName = getWalletName(matchingAccount!.walletId);
+
+      // Some login methods may require a single sign-on redirect, and take
+      // the user away from this page, so this promise might never resolve
+      const { openSessionRevalidation } = await import(
+        'views/modals/session_revalidation_modal'
+      );
+      const signerAddress: string = await new Promise((resolve, reject) => {
+        openSessionRevalidation({
+          walletAddress: address,
+          walletName,
+          walletSsoSource: matchingAccount.walletSsoSource,
+          canvasChainId,
+          onVerified: (verifiedAddress) => resolve(verifiedAddress),
+          onClose: () => {
+            const err = new Error();
+            (err as any).responseJSON = { error: 'Login canceled' };
+            reject(err);
+          },
+        });
+      });
+
+      // The user may have signed using a different account
+      const sessionReauthed = await controller.hasAuthenticatedSession(
+        canvasChainId,
+        address
+      );
+      if (!sessionReauthed) {
+        const err = new Error();
+        const { formatAddress } = await import(
+          'views/components/user/user_block'
+        );
+        (err as any).responseJSON = {
+          error: `Expected the address ${formatAddress(
+            address
+          )}, but this wallet has address ${formatAddress(signerAddress)}.`,
+        };
+        throw err;
+      }
```

Comment:
FYI @raykyri @jnaviask I am working right now on this piece. We have to move it out from controller to some reusable logic (custom hook probably), because
1. we should try to not add new logic to the controllers as we are trying to move away from controllers architecture
2. we should not handle UI from the controllers layer